Pure single assignments

keyboardDrummer opened the pull request:
### What was changed?
- Widen the rules for pure blocks. Allow multiple statements. Any statement but the last must be a variable declaration with an initializer. The last statement may now also be an if-else, besides a return.

### How has this been tested?
- Added files `PureMethodErrors` and `PureMethodVerification`

<small>By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache-2.0](https://github.com/aws/jverify?tab=Apache-2.0-1-ov-file#readme).</small>

keyboardDrummer committed:
Update documentation

keyboardDrummer committed:
Doc tweaks

keyboardDrummer committed:
Update docs

oyendrila-dobe commented on docs/guide/src/code_types.md:
```diff
@@ -26,48 +26,24 @@ We say code is pure if it does not have side effects and is deterministic (produ
+The rules for pure expressions are as follows:
 - Only methods annotated with `@Pure` can be called.
-- Local variables or fields can not be updated, so assignment operators such as `=` can not be used.
+- Assignment expression are not allowed
```
nitpik: should it be `is` or `expressions`?

oyendrila-dobe commented on docs/guide/src/code_types.md:
```diff
@@ -26,48 +26,24 @@ We say code is pure if it does not have side effects and is deterministic (produ
-Code that occurs in a method annotated with `@Pure` must follow those  rules, and additionally:
-- Only the following statements can be used:
-  - variable declaration statements,
-  - calls to `precondition`, `postcondition`, `reads`, and `check`
```
Is the possible list of allowed calls common knowledge or documented somewhere else in this doc?

keyboardDrummer replied:
Good question. Contracts are first introduced in the section `docs/guide/src/pre_and_postconditions.md`, where `precondition` and `postcondition` are introduced. There are more options (also more than the 4 that were previously mentioned here) and there are introduced in various sections.

oyendrila-dobe reviewed (approved):
Core change and test cases look good to me

keyboardDrummer committed:
Allow external contracts to specify a body for pure methods (#285)

### What was changed?
- Remove code that was needed for the old implementation of lambdas
- Make code that deals with external contracts more robust
- Allow external contracts to specify a body for pure methods

### How has this been tested?
- Add/update test files `LibraryContractGhostFieldVerification` and
`LibraryContractGhostFieldErrors`

<small>By submitting this pull request, I confirm that my contribution
is made under the terms of the
[Apache-2.0](https://github.com/aws/jverify?tab=Apache-2.0-1-ov-file#readme).</small>

---------

Co-authored-by: Robin Salkeld <salkeldr@amazon.com>

keyboardDrummer committed:
Merge branch 'externalPureBodyContracts' into pureSingleAssignments

keyboardDrummer committed:
Merge commit 'fd963e106032e4' into pureSingleAssignments

keyboardDrummer committed:
Fix oops

keyboardDrummer committed:
Fix oops
